src/main/golem/util/validation/dimensions.kt

```diff
@@ -0,0 +1,256 @@
+                    failed.add(Pair(context.matrixNames[index], matrix))
+                }
+            }
+        }
+        if (failed.size > 0) {
+            throw IndexOutOfBoundsException(concoctExceptionMessage(context, failed))
+        }
+    }
+
+
+    // Append text to the PrintStream, padded out to the given width with the given characters.
+    private fun PrintStream.writeColumn(width: Int, text: String, lpad: Char? = null, rpad: Char? = ' ',
+                                        gap: Char? = ' ') {
+        val nrpad = if (rpad == null) 0 else (width - text.length) / (if (lpad == null) 1 else 2)
+        val nlpad = if (lpad == null) 0 else (width - nrpad - text.length)
+        if (lpad != null)
+            for (x in 1..nlpad)
+                append(lpad)
+        append(text)
+        if (rpad != null)
+            for (x in 1..nrpad)
+                append(rpad)
+        if (gap != null)
+            append(gap)
+    }
+
+    // Build an english-description of numeric dimensions
+    private fun concoctFixedDescription(rows: String, cols: String): String {
+        if (rows == "1")
+            return "a ${cols}-vector"
+        if (cols == "1")
+            return "a ${rows}-vector"
+        return "${rows} by ${cols}"
+    }
+
+    // Build a list of things compareTo needs to be the same as. Looks up other things that have
+    // the same character-based dimension defined.
+    private fun concoctComparison(name: String, compareTo: String): List<String> {
+        val out = mutableListOf<String>()
+        if (compareTo !in seenInError) {
+            for ((name2, dims) in declared) {
+                val (rows, cols) = dims
+                if (name2 != name) {
+                    if (rows == compareTo) out.add("as ${name2} has rows")
+                    if (cols == compareTo) out.add("as ${name2} has columns")
+                }
+            }
+            seenInError.add(compareTo)
+        }
+        return out
+    }
+
+    // Build our pretty table of dimensions real vs actual.
+    private fun concoctExceptionMessage(context: ValidationContext, failed: List<Pair<String,
+    Matrix<Double>>>): String {
+        val strings = mutableListOf<String>()
+        for ((index, matrix) in context.matrices.withIndex()) {
+            val name = context.matrixNames[index]
+            val decl = declared[name]
+            if (decl != null) {
+                strings.add(name)
+                val (r, c) = decl
+                strings.add("${r}x${c}")
+                strings.add("${matrix.numRows()}x${matrix.numCols()}")
+            }
+        }
+        val headers = arrayOf("Matrix", "Required", "Actual")
+        val widths = mutableListOf(headers[0].length, headers[1].length, headers[2].length)
+        for ((index, str) in strings.withIndex())
+            if (str.length > widths[index % widths.size])
+                widths[index % widths.size] = str.length
+        val bstream = ByteArrayOutputStream()
```

Comment:
This code is actually based on the output code in Matrix.kt. Should I refactor to having a shared private function they both call that formats data in columns?